[ERROR] event.WriteError: event.newError: missing error: invalid parameter

**Describe the bug**

We get error `[ERROR] event.WriteError: event.newError: missing error: invalid parameter` on the Boundary workers when cancelling sessions in the UI.

```
Mar 10 16:48:12 ip-10-15-2-165 boundary[18049]: 2022-03-10T16:48:12.961Z [ERROR] event.WriteError: event.newError: missing error: invalid parameter
Mar 10 16:48:12 ip-10-15-2-165 boundary[18049]: 2022-03-10T16:48:12.961Z [ERROR] event.WriteError: unable to create new error to write error: <nil>
Mar 10 16:48:12 ip-10-15-2-165 boundary[18049]: {"id":"qRJGvP3Pw4","source":"https://hashicorp.com/boundary/us-east-2-worker-rxosqtnrrk","specversion":"1.0","type":"error","data":{"error":"failed to close WebSocket: failed to read frame header: EOF","error_fields":{},"id":"e_J4kGMpFMNu","version":"v0.1","op":"worker.(Worker).handleProxy","info":{"msg":"error closing client connection"}},"datacontentype":"application/cloudevents","time":"2022-03-10T16:48:12.963873011Z"}
```

**To Reproduce**
Steps to reproduce the behavior:
1. Under Targets tab, click Connect button on a target
2. Under Sessions tab, click Cancel on the session
3. See error in boundary worker log

**Expected behavior**
A clear and concise description of what you expected to happen.

**Additional context**
UI Version:
<img width="372" alt="image" src="https://user-images.githubusercontent.com/4521230/157713527-20424f57-99bc-4afb-963d-09e5783349a5.png">

Boundary version running on worker:
```
Mar 10 15:56:52 ip-10-15-2-165 boundary[18049]:             AWS KMS Region: us-east-2
Mar 10 15:56:52 ip-10-15-2-165 boundary[18049]:                        Cgo: disabled
Mar 10 15:56:52 ip-10-15-2-165 boundary[18049]:                 Listener 1: tcp (addr: "10.15.2.165:9202", max_request_duration: "1m30s", purpose: "proxy")
Mar 10 15:56:52 ip-10-15-2-165 boundary[18049]:                  Log Level: info
Mar 10 15:56:52 ip-10-15-2-165 boundary[18049]:                      Mlock: supported: true, enabled: true
Mar 10 15:56:52 ip-10-15-2-165 boundary[18049]:                    Version: Boundary v0.7.4
Mar 10 15:56:52 ip-10-15-2-165 boundary[18049]:                Version Sha: 221acff4cc4d1f9be7619a657274c043999e62cc
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -3,6 +3,10 @@
 Canonical reference for changes, improvements, and bugfixes for Boundary.
 
 ## Next
+### Bug Fixes
+* worker: create new error to prevent `event.newError: missing error: invalid parameter` and handle session cancel 
+with no TOFU token ([Issue](https://github.com/hashicorp/boundary/issues/1902),
+  [PR](https://github.com/hashicorp/boundary/pull/1929))
 ## 0.7.6 (2022/03/15)
 
 ### Bug Fixes

--- a/internal/servers/worker/handler.go
+++ b/internal/servers/worker/handler.go
@@ -138,6 +138,21 @@ func (w *Worker) handleProxy(listenerCfg *listenerutil.ListenerConfig) (http.Han
 			return
 		}
 
+		if handshake.Command == proxy.HANDSHAKECOMMAND_HANDSHAKECOMMAND_SESSION_CANCEL {
+			_, err := session.Cancel(ctx, sessClient, sessionId)
+			if err != nil {
+				event.WriteError(ctx, op, err, event.WithInfoMsg("unable to cancel session"))
+				if err = conn.Close(websocket.StatusInternalError, "unable to cancel session"); err != nil && !errors.Is(err, io.EOF) {
+					event.WriteError(ctx, op, err, event.WithInfoMsg("error closing client connection"))
+				}
+				return
+			}
+			if err = conn.Close(websocket.StatusNormalClosure, "session canceled"); err != nil && !errors.Is(err, io.EOF) {
+				event.WriteError(ctx, op, err, event.WithInfoMsg("error closing client connection"))
+			}
+			return
+		}
+
 		if tofuToken != "" {
 			if tofuToken != handshake.GetTofuToken() {
 				event.WriteError(ctx, op, errors.New("WARNING: mismatched tofu token"), event.WithInfo("session_id", sessionId))
@@ -148,7 +163,7 @@ func (w *Worker) handleProxy(listenerCfg *listenerutil.ListenerConfig) (http.Han
 			}
 		} else {
 			if sessStatus != pbs.SESSIONSTATUS_SESSIONSTATUS_PENDING {
-				event.WriteError(ctx, op, err, event.WithInfoMsg("no tofu token but not in correct session state"))
+				event.WriteError(ctx, op, errors.New("no tofu token but not in correct session state"), event.WithInfo("session_id", sessionId))
 				if err = conn.Close(websocket.StatusInternalError, "refusing to activate session"); err != nil {
 					event.WriteError(ctx, op, err, event.WithInfoMsg("error closing client connection"))
 				}
@@ -164,21 +179,6 @@ func (w *Worker) handleProxy(listenerCfg *listenerutil.ListenerConfig) (http.Han
 					return
 				}
 			}
-		}
-
-		if handshake.Command == proxy.HANDSHAKECOMMAND_HANDSHAKECOMMAND_SESSION_CANCEL {
-			_, err := session.Cancel(ctx, sessClient, sessionId)
-			if err != nil {
-				event.WriteError(ctx, op, err, event.WithInfoMsg("unable to cancel session"))
-				if err = conn.Close(websocket.StatusInternalError, "unable to cancel session"); err != nil && !errors.Is(err, io.EOF) {
-					event.WriteError(ctx, op, err, event.WithInfoMsg("error closing client connection"))
-				}
-				return
-			}
-			if err = conn.Close(websocket.StatusNormalClosure, "session canceled"); err != nil && !errors.Is(err, io.EOF) {
-				event.WriteError(ctx, op, err, event.WithInfoMsg("error closing client connection"))
-			}
-			return
 		}
 
 		// Verify the protocol has a supported proxy before calling AuthorizeConnection

